assimilation_code/modules/io/direct_netcdf_mod.f90

```diff
@@ -852,22 +851,41 @@ subroutine read_variables(ncfile_in, var_block, start_var, end_var, domain)
    var_size = get_variable_size(domain, i)
    iend = istart + var_size - 1
 
-   ! number of dimensions and length of each
-   allocate(dims(get_io_num_dims(domain, i)))
+   ! number of dimensions and length the variable
+   num_dims = get_io_num_dims(domain, i)
+   allocate(counts(num_dims))
+   allocate(slice_start(num_dims))
 
-   dims = get_io_dim_lengths(domain, i)
+   if (has_unlimited_dim(domain)) then
+
+      slice_start(:) = 1 ! default to read all dimensions start at 1
+      counts(num_dims) = 1 ! one slice of unlimited dimesion
+      counts(1:num_dims-1) = get_dim_lengths(domain, i) ! the state
```

Comment:
a couple of comments for the future here.  newer versions of netcdf allow for multiple unlimited dimensions so this approach will have to be altered if we run into a model that uses this in the input/output state to dart.  we can always require a model_to_dart and dart_to_model as a temporary measure once we do see how people might be using multiple unlimited dims.

also, it's possible that netcdf variables in files have additional (not-unlimited) dimensions that shouldn't be part of the state.  i can't remember which example or test had this, so for now it's not an issue.   but eventually some namelist items may be needed to address i/o issues - for example, the name(s) of the unlimited dimension(s) (e.g. Time), other name(s) of dimensions to ignore when reading into the state (e.g. leapfrog_step1 or something), and then for the ignored dimensions, what index should be used if the length is more than 1 (e.g. last, 1, etc).

Reply:
Totally agree with this, I'm keeping track State IO issues  here: https://github.com/NCAR/DART/issues/359 for a future refactor.